src/training_controller/prepare_training_logs.py

```diff
@@ -0,0 +1,176 @@
+# Standard Library
+import logging
+import os
+import shutil
+import subprocess
+import tarfile
+
+# Third Party
+import boto3
+import pandas as pd
+from botocore.client import Config
+
+logging.basicConfig(level=logging.INFO, format="%(asctime)s - %(name)s - %(message)s")
+MINIO_ACCESS_KEY = os.environ["MINIO_ACCESS_KEY"]
+MINIO_SECRET_KEY = os.environ["MINIO_SECRET_KEY"]
+
+
+class PrepareTrainingLogs:
+    def __init__(self, working_dir):
+        self.WORKING_DIR = working_dir
+        self.ES_DUMP_DIR = os.path.join(self.WORKING_DIR, "windows")
+        self.ES_DUMP_DIR_ZIPPED = self.ES_DUMP_DIR + ".tar.gz"
+        self.ES_DUMP_SAMPLE_LOGS_PATH = os.path.join(
+            self.WORKING_DIR, "sample_logs.json"
+        )
+
+        self.minio_client = boto3.resource(
+            "s3",
+            endpoint_url="http://minio.default.svc.cluster.local:9000",
+            aws_access_key_id=MINIO_ACCESS_KEY,
+            aws_secret_access_key=MINIO_SECRET_KEY,
+            config=Config(signature_version="s3v4"),
+        )
+        self.es_dump_data_path = ""
+
+    def save_window(self, window_start_time_ns, df):
+        df[["time_nanoseconds", "window_start_time_ns", "masked_log"]].to_json(
+            os.path.join(self.ES_DUMP_DIR, str(window_start_time_ns) + ".json.gz"),
+            orient="records",
+            lines=True,
+            compression="gzip",
+        )
+
+    def disk_size(self):
+        # Fetch size of disk
+        logging.info("Fetching size of the disk")
+        total, used, free = shutil.disk_usage("/")
+        logging.info("Disk Total: %d GiB" % (total // (2 ** 30)))
+        logging.info("Disk Used: %d GiB" % (used // (2 ** 30)))
+        logging.info("Disk Free: %d GiB" % (free // (2 ** 30)))
+        return free
+
+    def run_esdump(self, chunked_commands):
+        processes = set()
+        max_processes = 200
+
+        for index, chunk in enumerate(chunked_commands):
+            for query in chunk:
+                processes.add(subprocess.Popen(query))
+            for p in processes:
+                if p.poll() is None:
+                    p.wait()
+
+            logging.info("completed chunk {}".format(index))
+
+    def retrieve_sample_logs(self):
+        # Get the first 10k logs
+        logging.info("Retrieve sample logs from ES")
+        es_dump_cmd = (
+            'elasticdump --searchBody \'{"query": { "match_all": {} }, "_source": ["masked_log", "time_nanoseconds"]}\' --retryAttempts 10 --size=10000 --limit 10000 --input=http://elasticsearch-coordinating-only.default.svc.cluster.local:9200/logs --output=%s --type=data'
+            % self.ES_DUMP_SAMPLE_LOGS_PATH
```

Comment:
maybe get the latest logs? what is it by default?

Reply:
So by default it gets the first N logs from elasticdump.

Reply:
Getting the latest logs is possible too, can just make it sort in a descending order from the timestamp instead of ascending order as its currently being done.

Reply:
So with this query, since no sorting has been specified it will simply fetch the first 10000 logs that are present within Elasticsearch so it is not guaranteed that the timestamps of these logs would always be in ascending order. However, that logic could be added should we all decide on taking the first 10000 logs or the latest 10000 logs.

Reply:
get the latest 10000 logs

Reply:
get the latest 10000 logs